hanish/darksky.py

```diff
@@ -0,0 +1,192 @@
+# hanish.darksky
+# Handlers and functionality for making Dark Sky API requests.
+#
+# Author:   Benjamin Bengfort <benjamin@bengfort.com>
+# Created:  Mon May 08 12:04:11 2017 -0400
+#
+# Copyright (C) 2016 Bengfort.com
+# For license information, see LICENSE.txt
+#
+# ID: cdarksky.py [] benjamin@bengfort.com $
+
+"""
+Handlers and functionality for making Dark Sky API requests.
+"""
+
+##########################################################################
+## Imports and Module Vars
+##########################################################################
+
+import requests
+
+try:
+    # Python 3
+    from urllib.parse import urljoin
+except ImportError:
+    # Python 2
+    from urlparse import urljoin
+
+
+from datetime import date, datetime
+from expiringdict import ExpiringDict
+from .exceptions import HanishValueError
+from .exceptions import DarkSkyException
+
+
+DARKSKY_API_URL  = "https://api.darksky.net"
+API_CALLS_HEADER = "X-Forecast-API-Calls"
+
+
+##########################################################################
+## DarkSky API
+##########################################################################
+
+class DarkSky(object):
```

Comment:
Just another thought about providing multiple options for caching backends; and potentially this is what you meant in your question. In order to generalize the caching solution, in this case I would create an interface that supports `get` and `put` operations and is instantiated with max age and limit variables. The cache could then be generically instantiated with a variety of backends by passing a `cache_class` to the constructor of the API. 

In Python (I think Ruby as well), this is simply done by duck typing, but it would still probably be wise to create an abstract base class, `BaseCache` and manage our various backends via a class hierarchy. The `abc` module in Python will allow us to ensure the interface is correct, as well as to do type checking if needed. We could start simply by wrapping `ExpiringDict` in the base class or use multiple inheritance. For Java or Go, we would just define an interface and that's all that's required.